<!DOCTYPE html>
<html>

<head>
  <meta charset="UTF-8" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
<meta name="keywords" content="">
<meta name="description" content="有谱">
<meta name="theme-color" content="">
<title>这可能是我写过最全的GATK笔记 | 陈有朴</title>
<link rel="shortcut icon" href="/favicon.ico?v=1669425307641">
<link rel="stylesheet" href="/media/css/pisces.css">
<link rel="stylesheet" href="/media/fonts/font-awesome.css">
<link
  href="//fonts.loli.net/css?family=Monda:300,300italic,400,400italic,700,700italic|Roboto Slab:300,300italic,400,400italic,700,700italic|Rosario:300,300italic,400,400italic,700,700italic|PT Mono:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext"
  rel="stylesheet" type="text/css">

<link href="/media/hljs/styles/github.css"
  rel="stylesheet">

<link rel="stylesheet" href="/styles/main.css">
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>

<script src="/media/js/jquery.js"></script>
<script src="/media/hljs/highlight.js"></script>
<script src="https://cdn.jsdelivr.net/npm/velocity-animate@1.5.0/velocity.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/velocity-animate@1.5.0/velocity.ui.min.js"></script>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.css"
  integrity="sha384-zB1R0rpPzHqg7Kpt0Aljp8JPLqbXI3bhnPWROx27a9N0Ll6ZP/+DiW/UqRcLbRjq" crossorigin="anonymous">


<script defer src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.js"
  integrity="sha384-y23I5Q6l+B6vatafAwxRu/0oK/79VlbSz7Q9aiSZUvyWYIYsd+qj+o24G5ZU2zJz" crossorigin="anonymous"></script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/contrib/auto-render.min.js"
  integrity="sha384-kWPLUVMOks5AQFrykwIup5lo0m3iMkkHrD0uJ4H5cjeGihAutqP0yW0J6dpFiVkI" crossorigin="anonymous"
  onload="renderMathInElement(document.body);"></script>






  <meta name="description" content="这可能是我写过最全的GATK笔记" />
  <meta name="keywords" content="GATK,WGS" />
</head>

<body>
  <div class="head-top-line"></div>
  <div class="header-box">
    
<div class="pisces">
  <header class="header  ">
    <div class="blog-header box-shadow-wrapper bg-color " id="header">
      <div class="nav-toggle" id="nav_toggle">
        <div class="toggle-box">
          <div class="line line-top"></div>
          <div class="line line-center"></div>
          <div class="line line-bottom"></div>
        </div>
      </div>
      <div class="site-meta">       
        <div class="site-title">
          
            <a href="/" class="brand">
              <span>陈有朴</span>
            </a>  
          
        </div>
        
      </div>
      <nav class="site-nav" id="site_nav">
        <ul id="nav_ul">
          
            
            
            <li class="nav-item ">
              
              
                <a href="/" target="_self">
                  <i class="fa fa-globe"></i> 首页
                </a>
              
            </li>
          
            
            
            <li class="nav-item ">
              
              
                <a href="/archives" target="_self">
                  <i class="fa fa-globe"></i> Archives
                </a>
              
            </li>
          
            
            
            <li class="nav-item ">
              
              
                <a href="/tags" target="_self">
                  <i class="fa fa-globe"></i> Tags
                </a>
              
            </li>
          
            
            
            <li class="nav-item ">
              
              
                <a href="/post/about" target="_self">
                  <i class="fa fa-globe"></i> About Me
                </a>
              
            </li>
          
          
          
        </ul>
      </nav>
    </div>
  </header>
</div>

<script type="text/javascript"> 
 
  let showNav = true;

  let navToggle = document.querySelector('#nav_toggle'),
  siteNav = document.querySelector('#site_nav');
  
  function navClick() {
    let sideBar = document.querySelector('.sidebar');
    let navUl = document.querySelector('#nav_ul');
    navToggle.classList.toggle('nav-toggle-active');
    siteNav.classList.toggle('nav-menu-active');
    if (siteNav.classList.contains('nav-menu-active')) {
      siteNav.style = "height: " + (navUl.children.length * 42) +"px !important";
    } else {
      siteNav.style = "";
    }
  }

  navToggle.addEventListener('click',navClick);  
</script>
  </div>
  <div class="main-continer">
    
    <div
      class="section-layout pisces ">
      <div class="section-layout-wrapper">
        

<div class="sidebar">
  
    <div class="sidebar-box box-shadow-wrapper bg-color right-motion" id="sidebar">
      
      <div class="sidebar-body pisces" id="sidebar_body">
        
      </div>
    </div>
  
</div>
<script>
  const SIDEBAR_TITLE_ACTIVE = 'sidebar-title-active';
  const SIDEBAR_BODY_ACTIVE = 'sidebar-body-active';
  const SLIDE_UP_IN = 'slide-up-in';

  let sidebar = document.querySelector('#sidebar'),
  tocSideBar = document.querySelector('#tocSideBar'),
  metaSideBar = document.querySelector('#metaSideBar'),
  postToc = document.querySelector('#post_toc'),
  postSiteMeta = document.querySelector('#post_side_meta'),
  sidebarTitle = document.querySelector('.sidebar-title'),
  sidebarBody = document.querySelector('#sidebar_body');

  tocSideBar && tocSideBar.addEventListener('click', (e) => {
    toggleSidebar(e);
  });

  metaSideBar && metaSideBar.addEventListener('click', (e) => {
    toggleSidebar(e);
  });

  function toggleSidebar(e) {
    let currentTitle = document.querySelector("."+SIDEBAR_TITLE_ACTIVE);
    if (currentTitle == e.srcElement) {
      return ;
    }
    let current, showElement, hideElement;
    if (e.srcElement == metaSideBar) {
      showElement = postSiteMeta;
      hideElement = postToc;
    } else if (e.srcElement == tocSideBar){
      showElement = postToc;
      hideElement = postSiteMeta;
    }
    currentTitle.classList.remove(SIDEBAR_TITLE_ACTIVE);
    e.srcElement.classList.add(SIDEBAR_TITLE_ACTIVE);

    jQuery.Velocity(hideElement, 'stop');
    jQuery.Velocity(hideElement, 'transition.slideUpOut', {
      display: 'none',
      duration: 200,
      complete: function () {
        jQuery.Velocity(showElement, 'transition.slideDownIn', {
          duration: 200
        });
      }
    })
    hideElement.classList.remove(SIDEBAR_BODY_ACTIVE);
    showElement.classList.add(SIDEBAR_BODY_ACTIVE);
  }

  postToc && postToc.addEventListener('transitionend', function() {
    this.classList.remove(SLIDE_UP_IN);
  });

  if (sidebarBody) {
    if (sidebarBody.classList.contains('pisces') || sidebarBody.classList.contains('gemini')) {
      let hasFix = false;
      let scrollEl = document.querySelector('.main-continer');
      let limitTop = document.querySelector('#nav_ul').children.length * 42 + 162;
      window.addEventListener('scroll', function(e) {
        if (document.scrollingElement.scrollTop >= limitTop) {
          if (!hasFix) {
            sidebar.classList.add('sidebar-fixed');
            hasFix = true;
          }
        } else {
          if (hasFix) {
            sidebar.classList.remove('sidebar-fixed');
            hasFix = false;
          }
        }
      });
    }
  }
  
</script>
        <div class="section-box box-shadow-wrapper">
          <div class="section bg-color post post-page">
            <section class="post-header">
  <h1 class="post-title">
    <a class="post-title-link" href="https://Youpu-Chen.github.io/post/GATK-BESTPRACTICE/"> 这可能是我写过最全的GATK笔记 </a>
  </h1>
  <div class="post-meta">
    
    <span class="meta-item pc-show">
      <i class="fa fa-calendar-o"></i>
      <span class="language" data-lan="publish">发布于</span>
      <span class="publish-time" data-t="2022-08-11 11:23:01">2022-08-11</span>
      <span class="post-meta-divider pc-show">|</span>
    </span>
    
    <span class="meta-item">
      <i class="fa fa-folder-o"></i>
      <span class="pc-show language" data-lan="category-in">标签:</span>
       
      <a href="https://Youpu-Chen.github.io/tag/ZUP1ThQfW/"> <span>GATK</span> </a>、   
      <a href="https://Youpu-Chen.github.io/tag/aOqSaL_Yug/">
        <span>WGS</span>
      </a>
       
    </span>
    <span class="post-meta-divider">|</span>
    
    <span class="meta-item">
      <i class="fa fa-clock-o"></i>
      <span
        >17<span class="language" data-lan="minute"
          >分钟</span
        ></span
      >
    </span>
    <span class="meta-item">
      <span class="post-meta-divider">|</span>
      <i class="fa fa-file-word-o"></i>
      <span
        >3772<span class="pc-show language" data-lan="words"
          >字数</span
        ></span
      >
    </span>
    
  </div>
</section>

            <div class="post-body next-md-body" id="post_body">
              <h3 id="软件安装">软件安装</h3>
<p>官方网站：https://gatk.broadinstitute.org/hc/en-us</p>
<p>点击“Download GATK4”：<br>
<img src="https://Youpu-Chen.github.io//post-images/1660188520935.png" alt="" loading="lazy"></p>
<p>选择zip安装包：<br>
<img src="https://Youpu-Chen.github.io//post-images/1660188530491.png" alt="" loading="lazy"></p>
<pre><code class="language-bash">wget https://github.com/broadinstitute/gatk/releases/download/4.2.5.0/gatk-4.2.5.0.zip
unzip gatk-4.2.5.0.zip
</code></pre>
<h4 id="其他">其他</h4>
<p>安装GATK Best Practice所需要的软件：https://gatk.broadinstitute.org/hc/en-us/articles/360041320571--How-to-Install-all-software-packages-required-to-follow-the-GATK-Best-Practices</p>
<p>其他分析流程（bwa替代流程）：https://gatk.broadinstitute.org/hc/en-us/articles/4407897446939--How-to-Run-germline-single-sample-short-variant-discovery-in-DRAGEN-mode</p>
<h3 id="gatkjave选项">GATK：Jave选项</h3>
<p>GATK实际使用的命令为：<code>java -jar program.jar</code>，但是为了GATK的开发者为了方便将其添加到环境变量，对其进行了封装，即使用安装目录下的<strong>gatk可执行脚本</strong>可直接运行：<br>
<img src="https://Youpu-Chen.github.io//post-images/1660188563873.png" alt="" loading="lazy"></p>
<p>下面给出一个封装前后，设置额外参数的例子：</p>
<p>未进行封装前，设置额外参数的方式为<code>java -Xmx4G -jar gatk-package-4.2.2.0-local.jar</code></p>
<p>进行封装后，设置额外参数的方式为<code>gatk --java-options &quot;-Xmx4G&quot;</code></p>
<p>官方文档如下：</p>
<p>[1] https://gatk.broadinstitute.org/hc/en-us/articles/360035531892-GATK4-command-line-syntax</p>
<h3 id="gatkhallotypecaller">GATK：HallotypeCaller</h3>
<p>关于HallotypeCaller（GATK 3.6之后的版本）是否还需要使用重比对？</p>
<p>答案是不需要。</p>
<p>细节请看：</p>
<p>[1] https://github.com/broadinstitute/gatk-docs/blob/master/blog-2012-to-2019/2016-06-21-Changing_workflows_around_calling_SNPs_and_indels.md?id=7847</p>
<h3 id="gatkgenomicsdbimport">GATK：GenomicsDBImport</h3>
<p>在完成gatk HallotypeCaller分析这一步之后，可以选择GenomicsDBImport将生成的gvcf文件进行整合，便于后续的joint genotyping。</p>
<p>【标注】</p>
<ul>
<li>“GATK4 Best Practice for SNP and Indel”一般都选择GenomicsDBImport（而不是CombineGVCFs）进行gvcf文件的合并。GenomicsDBImport<strong>有一套独立的数据存储系统</strong>；</li>
<li>GenomicsDBImport与CombineGVCFs功能类似 —— 合并gvcf文件。前者将genomic loci作为划分依据（e.g. chromosome, scaffold, contig），后者使用sample；</li>
<li>可以使用SelectVariants，对GenomicsDBImport产生的数据库内容进行访问；</li>
</ul>
<h4 id="要求输入数据">要求输入数据</h4>
<p>GenomicsDBImport所要求的输入数据，为HallotypeCaller添加<code>-ERC GVCF</code>或<code>-ERC BP_RESOLUTION</code>参数生成的结果文件，即gvcf文件。</p>
<h4 id="参数设置强制要求">参数设置（强制要求）</h4>
<p>运行GenomicsDBImport的时候，有一些参数是必须的，还有一些参数是额外设置，可以用于增大文件读取速度或者进行个性化分析。</p>
<pre><code class="language-bash">--genomicsdb-workspace-path  # 构建GenomicsDatabase的目标文件夹
-V                           # gvcf文件名
--sample-name-map            # 一个包含所有gvcf ID的文本，使用tab分隔符，第一列为sample ID，第二列为gvcf ID
-L | --intervals             # 选择需要合并的基因组区域（每一行一个染色体编号）
</code></pre>
<p>额外参数设置：</p>
<pre><code class="language-bash">--batch-size        # 代表每批次能够读入多少个样本，默认为0，表示一次性全部读入。当样本数超过100时需要注意
</code></pre>
<h4 id="输入文件">输入文件</h4>
<p>GenomicsDBImport特殊的数据存储格式（e.g. ）</p>
<h4 id="genomicsdbimport示例代码">GenomicsDBImport：示例代码</h4>
<pre><code class="language-bash"># 每一个sample gvcf作为输入文件
gatk --java-options &quot;-Xmx4g -Xms4g&quot; GenomicsDBImport \
  -V data/gvcfs/mother.g.vcf.gz \
  -V data/gvcfs/father.g.vcf.gz \
  -V data/gvcfs/son.g.vcf.gz \
  --genomicsdb-workspace-path my_database \
  --tmp-dir=/path/to/large/tmp \
  -L 20

# 将所有sample gvcf名称放入map文件中
# 对应参数：--sample-name-map
gatk --java-options &quot;-Xmx4g -Xms4g&quot; \
   GenomicsDBImport \
   --genomicsdb-workspace-path my_database \
   --batch-size 50 \
   -L chr1:1000-10000 \
   --sample-name-map cohort.sample_map \
   --tmp-dir /path/to/large/tmp \
   --reader-threads 5

# 将新sample添加到GenomicsDBImport数据库中
# 对应参数：--genomicsdb-update-workspace-path
gatk --java-options &quot;-Xmx4g -Xms4g&quot; GenomicsDBImport \
  -V data/gvcfs/mother.g.vcf.gz \
  -V data/gvcfs/father.g.vcf.gz \
  -V data/gvcfs/son.g.vcf.gz \
  --genomicsdb-update-workspace-path my_database \
  --tmp-dir /path/to/large/tmp \
</code></pre>
<p>【标注】sample map文件使用tab分隔符，每一行一个gvcf文件名</p>
<pre><code class="language-bash">sample1      sample1.vcf.gz
sample2      sample2.vcf.gz
sample3      sample3.vcf.gz
</code></pre>
<p>还有一些细节就看看官方文档吧~</p>
<p>[1] https://gatk.broadinstitute.org/hc/en-us/articles/360057439331-GenomicsDBImport</p>
<p>[2] https://github.com/GenomicsDB/GenomicsDB/wiki</p>
<h3 id="gatkcombinegvcfs">GATK：CombineGVCFs</h3>
<p>CombineGVCFs的主要功能就是将<strong>HaplotypeCaller</strong>产生的gvcf文件，给合并。</p>
<p>【标注】</p>
<ul>
<li>有且只能是HaplotypeCaller产生的gvcf可以用作输入文件</li>
<li>VCF文件的合并，使用的是MergeVcfs，是Picard下的一个工具</li>
<li>官方给出的说明是：1000+以上的样本推荐使用GenomicsDBImport</li>
</ul>
<h4 id="combinegvcfs示例代码">CombineGVCFs：示例代码</h4>
<pre><code class="language-bash">gatk CombineGVCFs \
  -R reference.fasta \
  --variant sample1.g.vcf.gz \
  --variant sample2.g.vcf.gz \
  -O cohort.g.vcf.gz

# 将染色体拆分运行
gatk --java-options &quot;-Xmx4g -Xms4g&quot; CombineGVCFs -L 1 -V DRR083661.g.vcf.gz -O chr1.vcf.gz -R /home/chphl/2022_3_4_WGS/00.ref/arab_ref.fa
</code></pre>
<p>官方说明文档如下：</p>
<p>[1] https://gatk.broadinstitute.org/hc/en-us/articles/360037593911-CombineGVCFs</p>
<h3 id="gatkgenotypegvcfs">GATK：GenotypeGVCFs</h3>
<p>这边有一个非常关键词，“joint genotyping”。</p>
<p>genotyping，<strong>实际上就是发现给定群体（数据）中的DNA变异</strong>，包括SNP、INDEL、non-variation位点等。</p>
<ul>
<li>要求的输入数据是HaplotypeCaller附加“-ERC GVCF”或“-ERC BP_RESOLUTION”参数，所产生的gvcf文件</li>
<li>输出文件：VCF</li>
</ul>
<h4 id="genotypegvcfs示例代码">GenotypeGVCFs：示例代码</h4>
<pre><code class="language-bash">gatk --java-options &quot;-Xmx4g&quot; GenotypeGVCFs \
  -R Homo_sapiens_assembly38.fasta \
  -V input.g.vcf.gz \
  -O output.vcf.gz

# 当使用GenomicsDBImport作为合并HaplotypeCaller输出数据的工具时，
gatk GenotypeGVCFs \
    -R data/ref/ref.fasta \
    -V gendb://my_database \
    -O test_output.vcf
</code></pre>
<p>一些常用参数：</p>
<pre><code class="language-bash">--TMP_DIR                  # 使用暂时文件夹对结果文件进行保存
--max-genotype-count       # 每一个位置的genotype数量上限，默认为1024

# 其他参数
--sample-ploidy            # 在混池测序的时候需要注意，设置为“Number of samples in each pool * Sample Ploidy”
</code></pre>
<p>官方说明文档如下：</p>
<p>[1] https://gatk.broadinstitute.org/hc/en-us/articles/360037057852-GenotypeGVCFs</p>
<p>[2] https://gatk.broadinstitute.org/hc/en-us/articles/360035889971--How-to-Consolidate-GVCFs-for-joint-calling-with-GenotypeGVCFs</p>
<h3 id="gatkselectvariants">GATK：SelectVariants</h3>
<p>SelectVariants用于选择给定VCF文件中的个体 &amp; SNP类型（Select a subset of variants from a VCF file），功能可以概述为：</p>
<p>1、从给定VCF文件中，挑选个体（参数：-sn）</p>
<p>2、从给定VCF文件中，挑选对应区域（参数：--intervals）</p>
<p>3、选择不同类型的SNP（参数：--select-type）</p>
<p>e.g. 只挑选“INDELs”</p>
<p>一些常用参数：</p>
<pre><code class="language-bash">-R | --reference           # 给定参考基因组（官网说明文档给的是NULL，可有有无？）
--sn                       # 选择给定的样本，样本名保存在以.args结尾的文本中（一定要.args结尾！！！）
-xl-sn                     # 反向选择样本
--intervals                # 选择给定区域
--select-type-to-include | -select-type   # 选择保留给定类型的一种variants（注意，只有一种）
                                          # 没有给定的情况下，默认保留所有位点
                                          # ！！！多次设置来选择多种类型的位点
# 包括：INDEL, SNP, MIXED, MNP, SYMBOLIC, NO_VARIATION
# INDEL：insertion和deletion
# SNP：single nucleotide polymorphism
# MIXED：同一个位置上，不仅有SNP，还有IDEL（在多个个体水平）
# MNP：由相邻物理位置上组合成的SNP集合
# SYMBOLIC：
# NO_VARIATION：非多态性位点
 
--restrict-alleles-to        # 限制variants的类型为“ALL”，“BIALLELIC”，“MULTIALLELIC”其中的一种
--exclude-filtered           # 启用该FLAG值时，输出结果中只会包括有“PASS”标记的位点
--set-filtered-gt-to-nocall  # 默认情况下不开启，开启之后将“./.”位点全部过滤
</code></pre>
<p>其他参数：</p>
<pre><code class="language-bash">--selectExpressions     # 选择variants的标准（自行定义）
</code></pre>
<h4 id="需要注意的问题indexfeaturefile">需要注意的问题：IndexFeatureFile</h4>
<ul>
<li>在服务器中移动VCF文件时，转移到一个文件夹的时候忘记把对应VCF文件的<strong>索引</strong>也转移过来了</li>
<li>或者，本身就没有对该VCF创建索引</li>
</ul>
<p>可以使用如下命令，对目标VCF文件建立索引</p>
<pre><code class="language-bash">gatk --java-options &quot;-Xmx4G&quot; IndexFeatureFile --input input.vcf
</code></pre>
<h4 id="需要注意的问题createsequencedictionary">需要注意的问题：CreateSequenceDictionary</h4>
<p>在输入参考基因组（fasta格式）进行辅助分析的时候，不仅需要使用samtools对fasta文件进行fai索引的构建，还需要使用属于Picard工具包的CreateSequenceDictionary对其进行dict文件的构建：</p>
<pre><code class="language-bash">samtools faidx ref.fasta
# 结果文件：ref.fasta.fai

gatk CreateSequenceDictionary R=ref.fasta O=ref.dict
# 结果文件：ref.dict
</code></pre>
<p>官方说明文档：</p>
<p>[1] https://gatk.broadinstitute.org/hc/en-us/articles/360036362532-SelectVariants</p>
<p>[2] https://gatk.broadinstitute.org/hc/en-us/articles/360035530752-What-types-of-variants-can-GATK-tools-detect-or-handle-</p>
<p>[3] https://gatk.broadinstitute.org/hc/en-us/articles/360035890771-Biallelic-vs-Multiallelic-sites</p>
<p>[4] https://www.strand-ngs.com/files/manual/reference/snp.html</p>
<h3 id="gatkvariantfiltration">GATK：VariantFiltration</h3>
<p>VariantFiltration命令主要的功能是设置阈值，对SNP进行硬过滤，在FILTER这一列对位点进行标记。</p>
<ul>
<li>一般被过滤的位点，被标记为自行给定的FILTER名称</li>
<li>没有被过滤的位点，被标记为PASS</li>
</ul>
<h4 id="variantfiltration示例代码">VariantFiltration：示例代码</h4>
<pre><code class="language-bash">gatk VariantFiltration \
-R reference.fasta \
-V input.vcf.gz \
-O output.vcf.gz \
--filter-name &quot;my_filter1&quot; \
--filter-expression &quot;AB &lt; 0.2&quot; \
--filter-name &quot;my_filter2&quot; \
--filter-expression &quot;MQ0 &gt; 50&quot;
</code></pre>
<h4 id="一些参数的说明">一些参数的说明</h4>
<p>官方推荐将SNP和INDEL单独提取出来之后，再分别进行VariantFiltration。</p>
<p>【分析标注】SelectVariants使用“-select-type”对变异位点进行筛选时，只会根据给定类型，筛选单一子集（e.g. 给定选择SNP，就只会筛选SNP，不会选择SNP和INDEL共存的位点）。若需要选择混合类型的位点，使用参数“MIXED”</p>
<pre><code class="language-bash"># 示例hard-filtering
gatk VariantFiltration \
    -V snps.vcf.gz \
    -filter &quot;QD &lt; 2.0&quot; --filter-name &quot;QD2&quot; \
    -filter &quot;QUAL &lt; 30.0&quot; --filter-name &quot;QUAL30&quot; \
    -filter &quot;SOR &gt; 3.0&quot; --filter-name &quot;SOR3&quot; \
    -filter &quot;FS &gt; 60.0&quot; --filter-name &quot;FS60&quot; \
    -filter &quot;MQ &lt; 40.0&quot; --filter-name &quot;MQ40&quot; \
    -filter &quot;MQRankSum &lt; -12.5&quot; --filter-name &quot;MQRankSum-12.5&quot; \
    -filter &quot;ReadPosRankSum &lt; -8.0&quot; --filter-name &quot;ReadPosRankSum-8&quot; \
    -O snps_filtered.vcf.gz
gatk VariantFiltration \ 
    -V indels.vcf.gz \ 
    -filter &quot;QD &lt; 2.0&quot; --filter-name &quot;QD2&quot; \
    -filter &quot;QUAL &lt; 30.0&quot; --filter-name &quot;QUAL30&quot; \
    -filter &quot;FS &gt; 200.0&quot; --filter-name &quot;FS200&quot; \
    -filter &quot;ReadPosRankSum &lt; -20.0&quot; --filter-name &quot;ReadPosRankSum-20&quot; \ 
    -O indels_filtered.vcf.gz
</code></pre>
<p>再给出一篇文章中的过滤参数：</p>
<figure data-type="image" tabindex="1"><img src="https://Youpu-Chen.github.io//post-images/1660188613462.png" alt="" loading="lazy"></figure>
<p>官方说明文档如下：</p>
<p>[1] https://gatk.broadinstitute.org/hc/en-us/articles/360037434691-VariantFiltration</p>
<p>[2] https://gatk.broadinstitute.org/hc/en-us/articles/360035531112--How-to-Filter-variants-either-with-VQSR-or-by-hard-filtering【主要看“2. Hard filter a cohort callset with VariantFiltration”这部分就行】</p>
<h2 id="其他gatk命令">其他GATK命令</h2>
<h3 id="gatk碱基质量矫正">GATK：碱基质量矫正</h3>
<p>碱基质量矫正（Base Quality Score Recalibration/BQSR），这一步分析主要用于检测由于测序仪器造成的系统误差。</p>
<h4 id="bqsr理解">BQSR理解</h4>
<p>可以从以下2个方面对BQSR进行理解：</p>
<p>1、首先需要明确的一点是，BQSR是对碱基质量值进行矫正，而不是碱基。碱基质量值代表了该碱基的可信度（有百分之多少的概率，我们可以相信这个位点的测序情况是正确的，或者说有百分之多少的概率，认为这个位点是测错的），而base quality又与后续的SNP检测有关，因此对碱基质量值进行矫正非常关键。</p>
<p>【标注】也就是说，我们不能够决定一个低质量的A碱基，其原本是不是一个T，但是我们可以决定相信这个A的程度（基于base quality）</p>
<p>2、BQSR通过机器学习的方法对base quality进行矫正（本质都是<strong>回归</strong>）</p>
<p>【标注】</p>
<ul>
<li>作者给出的例子，当AA在序列中连着出现的时候，认为后续出现的碱基都增加了1%被测序的概率，因此对应的base quality也需要下降</li>
<li>上述影响是具有累加性的</li>
</ul>
<h4 id="bqsr分析过程">BQSR分析过程</h4>
<p>1、基于输入数据（bam），建立一个<strong>协变模型</strong>（这个名词有待商榷）</p>
<p>结果文件为一个recalibration file</p>
<p>2、使用ApplyBQSR对原始输入文件的base quality进行矫正，结果文件为bam文件（新生成），同时在上述步骤要求输入先验SNP位点。具体过程：</p>
<ul>
<li>统计全局差异（observed/reported quality vs expected/empirical quality）</li>
<li>每一个位点加上窗口计算的差异 + 每一个位点加上循环计算 &amp; dinucleotide差异</li>
</ul>
<p>【分析标注】</p>
<ul>
<li>“dinucleotide”【待解决】</li>
<li>推荐二次建立模型，看看recalibration对整体结果的影响变化。</li>
</ul>
<h4 id="bqsr示例代码">BQSR：示例代码</h4>
<pre><code class="language-bash"># 还没学，用不到
</code></pre>
<p>官方文档如下：</p>
<p>[1] https://gatk.broadinstitute.org/hc/en-us/articles/360035890531-Base-Quality-Score-Recalibration-BQSR-</p>
<p>[2] https://gatk.broadinstitute.org/hc/en-us/articles/360036898312-BaseRecalibrator</p>
<h3 id="gatkvariantrecalibrator-applyvqsr">GATK：VariantRecalibrator &amp; ApplyVQSR</h3>
<p>VariantRecalibrator采用机器学习的方式，计算位点的VQSLOD值（取代该位点原本的QUAL），作为后续分析的指标，一般分为以下2个步骤：</p>
<ul>
<li>使用高质量的VCF数据集，训练模型</li>
<li>将模型应用到我们自己的数据当中去</li>
</ul>
<p>从上述的话，其实也能很明显地看出来，VariantRecalibrator &amp; ApplyVQSR这套方法的缺陷 —— 需要先验的高质量VCF数据集。</p>
<p>输入数据的要求：</p>
<figure data-type="image" tabindex="2"><img src="https://Youpu-Chen.github.io//post-images/1660188666242.png" alt="" loading="lazy"></figure>
<h4 id="variantrecalibrator示例运行">VariantRecalibrator：示例运行</h4>
<pre><code class="language-bash"># exome data
gatk VariantRecalibrator \
  -R Homo_sapiens_assembly38.fasta \
  -V input.vcf.gz \
  --resource hapmap,known=false,training=true,truth=true,prior=15.0:hapmap_3.3.hg38.sites.vcf.gz \
  --resource omni,known=false,training=true,truth=false,prior=12.0:1000G_omni2.5.hg38.sites.vcf.gz \
  --resource 1000G,known=false,training=true,truth=false,prior=10.0:1000G_phase1.snps.high_confidence.hg38.vcf.gz \
  --resource dbsnp,known=true,training=false,truth=false,prior=2.0:Homo_sapiens_assembly38.dbsnp138.vcf.gz \
  -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \
  -mode SNP \
  -O output.recal \
  --tranches-file output.tranches \
  --rscript-file output.plots.R
  
# Allele-specific version of the SNP recalibration
gatk VariantRecalibrator \
  -R Homo_sapiens_assembly38.fasta \
  -V input.vcf.gz \
  -AS \
  --resource hapmap,known=false,training=true,truth=true,prior=15.0:hapmap_3.3.hg38.sites.vcf.gz \
  --resource omni,known=false,training=true,truth=false,prior=12.0:1000G_omni2.5.hg38.sites.vcf.gz \
  --resource 1000G,known=false,training=true,truth=false,prior=10.0:1000G_phase1.snps.high_confidence.hg38.vcf.gz \
  --resource dbsnp,known=true,training=false,truth=false,prior=2.0:Homo_sapiens_assembly38.dbsnp138.vcf.gz \
  -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \
  -mode SNP \
  -O output.AS.recal \
  --tranches-file output.AS.tranches \
  --rscript-file output.plots.AS.R
</code></pre>
<p>官方说明文档如下：</p>
<p>[1] https://gatk.broadinstitute.org/hc/en-us/articles/360036510892-VariantRecalibrator</p>
<p>[2] https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-</p>
<h3 id="gatkmergevcfs">GATK：MergeVCFs</h3>
<p>MergeVCFs是包含在Picard内的一个工具，用于合并不同区域的VCF文件</p>
<p>对输入文件的要求如下：</p>
<ul>
<li>每个VCF文件所包含的sample要一致</li>
<li>Input file headers must be contain compatible declarations for common annotations (INFO, FORMAT fields) and filters，即一些通用信息要包含（e.g. INFO, FORMAT, filters）</li>
<li>每个VCF文件包含的SNP，要求经过排序</li>
</ul>
<h4 id="mergevcfs示例代码">MergeVCFs：示例代码</h4>
<pre><code class="language-bash">java -jar /path/picard.jar MergeVcfs \
          I=input_variants.01.vcf \
          I=input_variants.02.vcf.gz \
          O=output_variants.vcf.gz
</code></pre>
<h2 id="题外话">题外话</h2>
<h3 id="gatk什么是read-group-如何添加read-group信息">GATK：什么是read group &amp; 如何添加read group信息</h3>
<p>在进行bwa mem比对时，需要向bam/sam文件中，添加read group信息。</p>
<p>如果bam/sam文件中，没有对应信息，则会报错。</p>
<h4 id="1gatk需要的read-group信息是什么">（1）GATK需要的read group信息是什么？</h4>
<pre><code class="language-bash">ID = Read group identifier
# 每一个read group 独有的ID，每一对reads 均有一个独特的ID，可以自定义命名；

PL = Platform
# 测序平台：ILLUMINA, SOLID, LS454, HELICOS and PACBIO，不区分大小写；

SM = sample
# reads属于的样品名；SM要设定正确，因为GATK产生的VCF文件也使用这个名字;

LB = DNA preparation library identifier
# 对一个read group的reads进行重复序列标记时，需要使用LB来区分reads来自那条lane;有时候，同一个库可能在不同的lane上完成测序;为了加以区分，
# 同一个或不同库只要是在不同的lane产生的reads都要单独给一个ID. 一般无特殊说明，成对儿read属于同一库，可自定义，比如：library1
</code></pre>
<h4 id="2如何进行read-group信息添加">（2）如何进行read group信息添加？</h4>
<p>图示：<br>
<img src="https://Youpu-Chen.github.io//post-images/1660299121752.png" alt="" loading="lazy"></p>
<pre><code class="language-bash">@A00312:194:HFK5KDSX2  # 测序仪器
4                      # lane ID
1101                   # tail坐标
24469                  # tail x坐标
13150                  # tail y坐标
# bwa比对时，对read group信息进行添加
bwa -R '@RG\tID:lane_${lane}\tPL:illumina\tSM:${sample}'
</code></pre>
<p>参考资料如下：</p>
<p>https://www.jianshu.com/p/c41e8f3266b4</p>
<h2 id="写在最后">写在最后</h2>
<p>变异分析流程的软件有很多，金标准只有GATK一个（针对于WGS和WES），但是GATK自己本身就不断在更新。所以说，只学习别人软件的用法，这样还是走不长远的，我觉得还是需要以后自己有能力去做一款软件，靠这个吃饭，我觉得比较ok。</p>

            </div>
            
            
              <div class="post-footer">
  <ul class="post-copyright">
    <li class="post-copyright-author">
      <strong class="language" data-lan="author">本文作者：</strong>
      
    </li>
    <li class="post-copyright-link">
      <strong class="language" data-lan="link">本文链接：</strong>
      <a href="https://Youpu-Chen.github.io/post/GATK-BESTPRACTICE/" title="这可能是我写过最全的GATK笔记">https://Youpu-Chen.github.io/post/GATK-BESTPRACTICE/</a>
    </li>
    <li class="post-copyright-license">
      <strong class="language" data-lan="copyright">版权声明： </strong>
      本博客所有文章除特别声明外，均采用<a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" rel="noopener" target="_blank"><i class="fa fa-fw fa-creative-commons"></i> BY-NC-SA</a> 许可协议。转载请注明出处！
    </li>
  </ul>
  <div class="tags">
    
      <a href="https://Youpu-Chen.github.io/tag/ZUP1ThQfW/"># GATK</a>
    
      <a href="https://Youpu-Chen.github.io/tag/aOqSaL_Yug/"># WGS</a>
    
  </div>
  <div class="nav">
    <div class="nav-prev">
      
        <i class="fa fa-chevron-left"></i>
        <a class="nav-pc-next" title="写给二十二岁的自己" href="https://Youpu-Chen.github.io/post/tomy22/">写给二十二岁的自己</a class="nav-pc-next">
        <a class="nav-mobile-prev" title="写给二十二岁的自己" href="https://Youpu-Chen.github.io/post/tomy22/">上一篇</a>
      
    </div>
    <div class="nav-next">
      
        <a class="nav-pc-next" title="「Hello Gridea」Here comes Youpu-Chen" href="https://Youpu-Chen.github.io/post/hello-gridea/">「Hello Gridea」Here comes Youpu-Chen</a>
        <a class="nav-mobile-next" title="「Hello Gridea」Here comes Youpu-Chen" href="https://Youpu-Chen.github.io/post/hello-gridea/">下一篇</a>
        <i class="fa fa-chevron-right"></i>
      
    </div>
  </div>
</div>
            
            
  
    
    
      <link rel="stylesheet" href="https://unpkg.com/disqusjs@1.1/dist/disqusjs.css">
<script src="https://unpkg.com/disqusjs@1.1/dist/disqus.js"></script>

<div id="disqus_thread"></div>

<script>
  var options = {
    shortname: 'DisqusComments',
    apikey: '0D4eig7RqYsljZD0i62dy7y5H2AVm2gE3T9qwx1S16zyJejdxXypcpTtxHm07Eig',
  }
  if ('https://disqus.com/api/oauth/2.0/authorize/') {
    options.api = 'https://disqus.com/api/oauth/2.0/authorize/'
  }
  var dsqjs = new DisqusJS(options)
</script>
    
  

          </div>
        </div>
      </div>
    </div>
    <div class="footer-box">
  <footer class="footer">
    <center id="runTimeBox">
      已运行:<span id="run_time"></span>
    </center>
    <span id="busuanzi_container_site_pv">浏览数:<span id="busuanzi_value_site_pv"></span> 次</span>
    <span class="post-meta-divider">|</span>
    <span id="busuanzi_container_site_uv">访客数:<span id="busuanzi_value_site_uv"></span> 人</span>

    <script>
      BirthDay = new Date('');
      if (BirthDay.getTime()) {
        function runTime() {
          str = "";
          today = new Date();
          timeold = today.getTime() - BirthDay.getTime();
          msPerDay = 24 * 60 * 60 * 1000;
          e_daysold = timeold / msPerDay;
          daysold = Math.floor(e_daysold);
          str += daysold + "天";
          return str;
        }
        setInterval(function () {
          $("#run_time").html(runTime());
        }, 1000);
      } else {
        document.querySelector('.footer').removeChild(document.querySelector('#runTimeBox'));
      }
    </script>
    <div class="poweredby">
      Powered by <a href="https://github.com/getgridea/gridea" target="_blank"> Gridea</a>
    </div>
  </footer>
  
    
        <div class="pisces back-to-top" id="back_to_top">
          <i class="fa fa-arrow-up"></i>
          
            <span class="scrollpercent"> <span id="back_to_top_text">0</span>% </span>
            
        </div>
        
          
            
                
                  
                        
                          

                            
</div>
<script>
  let sideBarOpen = "sidebar-open";
  let body = document.body;
  let back2Top = document.querySelector("#back_to_top"),
    back2TopText = document.querySelector("#back_to_top_text"),
    drawerBox = document.querySelector("#drawer_box"),
    rightSideBar = document.querySelector(".sidebar"),
    viewport = document.querySelector("body");

  function scrollAnimation(currentY, targetY) {
    let needScrollTop = targetY - currentY;
    let _currentY = currentY;
    setTimeout(() => {
      const dist = Math.ceil(needScrollTop / 10);
      _currentY += dist;
      window.scrollTo(_currentY, currentY);
      if (needScrollTop > 10 || needScrollTop < -10) {
        scrollAnimation(_currentY, targetY);
      } else {
        window.scrollTo(_currentY, targetY);
      }
    }, 1);
  }

  back2Top.addEventListener("click", function (e) {
    scrollAnimation(document.scrollingElement.scrollTop, 0);
    e.stopPropagation();
    return false;
  });

  window.addEventListener("scroll", function (e) {
    let percent =
      (document.scrollingElement.scrollTop /
        (document.scrollingElement.scrollHeight -
          document.scrollingElement.clientHeight)) *
      100;
    if (percent > 1 && !back2Top.classList.contains("back-top-active")) {
      back2Top.classList.add("back-top-active");
    }
    if (percent == 0) {
      back2Top.classList.remove("back-top-active");
    }
    if (back2TopText) {
      back2TopText.textContent = Math.floor(percent);
    }
  });

  let hasCacu = false;
  window.addEventListener("resize", function (e) {
    calcuHeight();
  });

  function calcuHeight() {
    // 动态调整站点概览高度
    if (
      (!hasCacu && back2Top.classList.contains("pisces")) ||
      back2Top.classList.contains("gemini")
    ) {
      let sideBar = document.querySelector(".sidebar");
      let navUl = document.querySelector("#site_nav");
      sideBar.style =
        "margin-top:" + (navUl.offsetHeight + navUl.offsetTop + 15) + "px;";
      hasCacu = true;
    }
  }
  calcuHeight();

  let open = false,
    MOTION_TIME = 300,
    RIGHT_MOVE_DIS = "320px";

  if (drawerBox) {
    let rightMotions = document.querySelectorAll(".right-motion");
    let right = drawerBox.classList.contains("right");

    let transitionDir = right
      ? "transition.slideRightIn"
      : "transition.slideLeftIn";

    let openProp, closeProp;
    if (right) {
      openProp = {
        paddingRight: RIGHT_MOVE_DIS,
      };
      closeProp = {
        paddingRight: "0px",
      };
    } else {
      openProp = {
        paddingLeft: RIGHT_MOVE_DIS,
      };
      closeProp = {
        paddingLeft: "0px",
      };
    }

    drawerBox.onclick = function () {
      open = !open;
      jQuery.Velocity(rightSideBar, "stop");
      jQuery.Velocity(viewport, "stop");
      jQuery.Velocity(rightMotions, "stop");
      if (open) {
        jQuery.Velocity(
          rightSideBar,
          {
            width: RIGHT_MOVE_DIS,
          },
          {
            duration: MOTION_TIME,
            begin: function () {
              jQuery.Velocity(rightMotions, transitionDir, {});
            },
          }
        );
        jQuery.Velocity(viewport, openProp, {
          duration: MOTION_TIME,
        });
      } else {
        jQuery.Velocity(
          rightSideBar,
          {
            width: "0px",
          },
          {
            duration: MOTION_TIME,
            begin: function () {
              jQuery.Velocity(rightMotions, {
                opacity: 0,
              });
            },
          }
        );
        jQuery.Velocity(viewport, closeProp, {
          duration: MOTION_TIME,
        });
      }
      for (let i = 0; i < drawerBox.children.length; i++) {
        drawerBox.children[i].classList.toggle("muse-line");
      }
      drawerBox.classList.toggle(sideBarOpen);
    };
  }

  // 链接跳转
  let newWindow = "";
  if (newWindow === "true") {
    let links = document.querySelectorAll(".post-body a");
    links.forEach((item) => {
      if (!item.classList.contains("btn")) {
        item.setAttribute("target", "_blank");
      }
    });
  }

  let faSearch = document.querySelector("#fa_search");
  faSearch &&
    faSearch.addEventListener("click", function () {
      document.querySelector("#search_mask").style = "";
    });

  // 代码高亮
  hljs.initHighlightingOnLoad();

  // 离开当前页title变化
  var leaveTitle = "";
  var normal_title = document.title;
  if (leaveTitle) {
    document.addEventListener("visibilitychange", function () {
      if (document.visibilityState == "hidden") {
        normal_title = document.title;
        document.title = leaveTitle;
      } else {
        document.title = normal_title;
      }
    });
  }
</script>

<link rel="stylesheet" href="/media/css/jquery.fancybox.css" />
<script src="/media/js/jquery.fancybox.js"></script>

<script>
  let images = document.querySelectorAll(".section img");
  images.forEach((image) => {
    var parent = image.parentElement;
    var next = image.nextElementSibling;
    parent.removeChild(image);
    var aelem = document.createElement("a");
    aelem.href = image.src;
    aelem.dataset["fancybox"] = "images";
    aelem.dataset["rel"] = "fancybox-button";
    aelem.classList.add("fancybox");
    aelem.appendChild(image);
    parent.insertBefore(aelem, next);
  });
</script>
    <div class="reward-mask" style="display: none;">
  <div class="reward-relative">
    <span class="close" aria-hidden="true">x</span>
    <div class="reward-body">
      <h2>感谢您的支持，我会继续努力的!</h2>
      <div class="reward-img-box">
        <div style="position: relative; width: 140px; height: 140px;">
          
          
          
        </div>
      </div>
      <p class="reward-word"></p>
      <p class="reward-tip"></p>
    </div>
    <div class="bottom">
      
      
    </div>
  </div>
</div>
<style>
  .reward-mask {
    position: fixed;
    z-index: 99999;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: #00000054;
  }

  .reward-relative {
    position: relative;
    width: 480px;
    text-align: center;
    margin: 0 auto;
    border-radius: 5px;
    background-color: #fff;
    top: 50%;
    margin-top: -205px;
  }

  .reward-relative .close {
    position: absolute;
    right: 10px;
    font-weight: normal;
    font-size: 16px;
    color: #929292;
  }

  .reward-body {
    padding: 40px 20px 20px;
  }

  .bottom {
    display: flex;
  }

  .reward-btn {
    text-align: center;
  }

  .reward-btn-text {
    display: inline-block;
    cursor: pointer;
    width: 60px;
    height: 60px;
    line-height: 60px;
    border-radius: 50%;
    background-color: #ff9734;
    color: #FFF;
    margin-top: 20px;
  }

  .pay-text {
    margin-top: 10px;
    padding: 10px;
    flex: 1;
    transition: all .2s linear;
  }

  .pay-text:hover {
    background-color: #a5a5a536;
  }

  .reward-body h2 {
    padding-top: 10px;
    text-align: center;
    color: #a3a3a3;
    font-size: 16px;
    font-weight: normal;
    margin: 0 0 20px;
  }

  .reward-body h2:after,
  .reward-body h2:before {
    font-family: Arial, Helvetica, sans-serif;
    background: 0 0;
    width: 0;
    height: 0;
    font-style: normal;
    color: #eee;
    font-size: 80px;
    position: absolute;
    top: 20px;
  }

  .reward-body h2:before {
    content: '\201c';
    left: 50px;
  }

  .reward-body h2:after {
    content: '\201d';
    right: 80px;
  }

  .reward-img-box {
    display: inline-block;
    padding: 10px;
    border: 6px solid #ea5f00;
    margin: 0 auto;
    border-radius: 3px;
    position: relative;
  }

  .reward-img {
    position: absolute;
    left: 0px;
    top: 0px;
    width: 100%;
    height: 100%;
  }

  @media (max-width: 767px) {
    .reward-relative {
      height: 100%;
      top: 0px;
      margin-top: 0;
      width: auto;
    }

    .reward-relative .bottom {
      flex-direction: column;
    }

    .reward-relative .pay-text {
      width: 80%;
      margin: 5px auto;
      border: 1px solid silver;
      padding: 6px;
      border-radius: 4px;
    }

    .reward-body h2:after {
      right: 40px;
    }

    .reward-body h2:after,
    .reward-body h2:before {
      font-size: 60px;
    }

    .reward-body h2:before {
      left: 20px;
    }
  }
</style>
<script>
  !function () {
    var mask = document.querySelector('.reward-mask');
    let close = document.querySelector('.reward-relative .close');
    let rewardBtn = document.querySelector('.reward-btn');

    let zfb = document.querySelector('#zfb'),
      wx = document.querySelector('#wx'),
      zfbBtn = document.querySelector('#zfbBtn'),
      wxBtn = document.querySelector('#wxBtn');

    if (zfbBtn && wxBtn) {
      zfbBtn.addEventListener('click', () => {
        jQuery.Velocity(zfb, 'transition.slideLeftIn', {
          duration: 400
        });
        jQuery.Velocity(wx, 'transition.slideRightOut', {
          display: 'none',
          duration: 400
        });
      });

      wxBtn.addEventListener('click', () => {
        jQuery.Velocity(wx, 'transition.slideRightIn', {
          duration: 400
        });
        jQuery.Velocity(zfb, 'transition.slideLeftOut', {
          display: 'none',
          duration: 400
        });
      });
    }

    rewardBtn.addEventListener('click', (e) => {
      jQuery.Velocity(mask, 'transition.slideDownIn', {
        duration: 400
      })
    });

    close.addEventListener('click', (e) => {
      e.preventDefault();
      jQuery.Velocity(mask, 'transition.slideUpOut', {
        duration: 400
      })
    })
  }()
</script>

  </div>
</body>

<input hidden id="copy" />
<script>
  !function () {
    let times = document.querySelectorAll('.publish-time');
    for (let i = 0; i < times.length; i++) {
      let date = times[i].dataset.t;
      let time = Math.floor((new Date().getTime() - new Date(date).getTime()) / 1000);
      if (time < 60) {
        str = time + '秒之前';
      } else if (time < 3600) {
        str = Math.floor(time / 60) + '分钟之前';
      } else if (time >= 3600 && time < 86400) {
        str = Math.floor(time / 3600) + '小时之前';
      } else if (time >= 86400 && time < 259200) {
        str = Math.floor(time / 86400) + '天之前';
      } else {
        str = times[i].textContent;
      }
      times[i].textContent = str;
    }
  }();
</script>

<script>
  let language = '';
  if (language !== '') {
    let map = new Map();
    if (language === 'en') {
      map.set('search', 'Search');
      map.set('category', 'Categories');
      map.set('article', 'Articles');
      map.set('tag', 'Tags');
      map.set('top', 'Top');
      map.set('publish', 'published');
      map.set('minute', ' minutes');
      map.set('read-more', 'Read More');
      map.set('view', 'View');
      map.set('words', ' words');
      map.set('category-in', 'category in');
      map.set('preview', 'Meta');
      map.set('index', 'Toc');
      map.set('no-archives', "You haven't created yet");
      map.set('archives', " articles in total");
      map.set('cloud-tags', " tags in total");
      map.set('copyright', "Copyright: ");
      map.set('author', "Author: ");
      map.set('link', "Link: ");
      map.set('leave-message', "Leave a message");
      map.set('format', "Links Format");
      map.set('site-name', "Name: ");
      map.set('site-link', "Link: ");
      map.set('site-desc', "Desc: ");
      map.set('stat', " related results, taking ");
      map.set('stat-time', " ms");
      map.set('site-img', "Image: ");
    }

    if (map.size > 0) {
      let lanElems = document.querySelectorAll('.language');
      lanElems.forEach(elem => {
        let lan = elem.dataset.lan, text = map.get(lan);
        if (elem.__proto__ === HTMLInputElement.prototype) {
          elem.placeholder = text
        } else {
          if (elem.dataset.count) {
            text = elem.dataset.count + text;
          }
          elem.textContent = text;
        }
      })
    }
  }

  window.Clipboard = (function (window, document, navigator) {
    var textArea,
      copy;

    // 判断是不是ios端
    function isOS() {
      return navigator.userAgent.match(/ipad|iphone/i);
    }
    //创建文本元素
    function createTextArea(text) {
      textArea = document.createElement('textArea');
      textArea.value = text;
      textArea.style.width = 0;
      textArea.style.height = 0;
      textArea.clientHeight = 0;
      textArea.clientWidth = 0;
      document.body.appendChild(textArea);
    }
    //选择内容
    function selectText() {
      var range,
        selection;

      if (isOS()) {
        range = document.createRange();
        range.selectNodeContents(textArea);
        selection = window.getSelection();
        selection.removeAllRanges();
        selection.addRange(range);
        textArea.setSelectionRange(0, 999999);
      } else {
        textArea.select();
      }
    }

    //复制到剪贴板
    function copyToClipboard() {
      try {
        document.execCommand("Copy")
      } catch (err) {
        alert("复制错误！请手动复制！")
      }
      document.body.removeChild(textArea);
    }

    copy = function (text) {
      createTextArea(text);
      selectText();
      copyToClipboard();
    };

    return {
      copy: copy
    };
  })(window, document, navigator);

  function copyCode(e) {
    if (e.srcElement.tagName === 'SPAN' && e.srcElement.classList.contains('copy-code')) {
      let code = e.currentTarget.querySelector('code');
      var text = code.innerText;
      if (e.srcElement.textContent === '复制成功') {
        return;
      }
      e.srcElement.textContent = '复制成功';
      (function (elem) {
        setTimeout(() => {
          if (elem.textContent === '复制成功') {
            elem.textContent = '复制代码'
          }
        }, 1000);
      })(e.srcElement)
      Clipboard.copy(text);
    }
  }

  let pres = document.querySelectorAll('pre');
  pres.forEach(pre => {
    let code = pre.querySelector('code');
    let copyElem = document.createElement('span');
    copyElem.classList.add('copy-code');
    copyElem.textContent = '复制代码';
    pre.appendChild(copyElem);
    pre.onclick = copyCode
  })

</script>
<script src="/media/js/motion.js"></script>


  <script
    src="https://cdn.jsdelivr.net/gh/cferdinandi/smooth-scroll/dist/smooth-scroll.polyfills.min.js"></script>
  <script>
    var scroll = new SmoothScroll('a[href*="#"]', {
      speed: 200
    });
  </script>






</html>